browser/net/brave_proxying_url_loader_factory.cc

```diff
@@ -0,0 +1,969 @@
+        .Run(net::ERR_ABORTED, base::nullopt);
+  }
+  if (on_headers_received_callback_) {
+    std::move(on_headers_received_callback_)
+        .Run(net::ERR_ABORTED, base::nullopt, GURL());
+  }
+  if (ctx_) {
+    factory_->request_handler_->OnURLRequestDestroyed(ctx_);
+  }
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::Restart() {
+  UpdateRequestInfo();
+  RestartInternal();
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::UpdateRequestInfo() {
+  // TODO(iefremov): Update |ctx_| here and get rid of multiple spots where
+  // it is refilled.
+  current_request_uses_header_client_ =
+      factory_->url_loader_header_client_binding_ &&
+      request_.url.SchemeIsHTTPOrHTTPS() && network_service_request_id_ != 0;
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::RestartInternal() {
+  request_completed_ = false;
+
+  // If the header client will be used, we start the request immediately, and
+  // OnBeforeSendHeaders and OnSendHeaders will be handled there. Otherwise,
+  // send these events before the request starts.
+  base::RepeatingCallback<void(int)> continuation;
+  if (current_request_uses_header_client_) {
+    continuation = base::BindRepeating(
+        &InProgressRequest::ContinueToStartRequest, weak_factory_.GetWeakPtr());
+  } else {
+    continuation =
+        base::BindRepeating(&InProgressRequest::ContinueToBeforeSendHeaders,
+                            weak_factory_.GetWeakPtr());
+  }
+  redirect_url_ = GURL();
+  ctx_ = std::make_shared<brave::BraveRequestInfo>();
+  brave::BraveRequestInfo::FillCTX(request_, render_process_id_,
+                                   frame_tree_node_id_, request_id_,
+                                   resource_context_, ctx_);
+  int result = factory_->request_handler_->OnBeforeURLRequest(
+      ctx_, continuation, &redirect_url_);
+
+  if (result == net::ERR_BLOCKED_BY_CLIENT) {
+    // The request was cancelled synchronously. Dispatch an error notification
+    // and terminate the request.
+    network::URLLoaderCompletionStatus status(result);
+    OnRequestError(status);
+    return;
+  }
+
+  if (result == net::ERR_IO_PENDING) {
+    // One or more listeners is blocking, so the request must be paused until
+    // they respond. |continuation| above will be invoked asynchronously to
+    // continue or cancel the request.
+    //
+    // We pause the binding here to prevent further client message processing.
+    if (proxied_client_binding_.is_bound())
+      proxied_client_binding_.PauseIncomingMethodCallProcessing();
+
+    // Pause the header client, since we want to wait until OnBeforeRequest has
+    // finished before processing any future events.
+    if (header_client_binding_)
+      header_client_binding_.PauseIncomingMethodCallProcessing();
+    return;
+  }
+  DCHECK_EQ(net::OK, result);
+
+  continuation.Run(net::OK);
+}
+
+void BraveProxyingURLLoaderFactory::InProgressRequest::FollowRedirect(
```

Comment:
can we subclass WebRequestProxyingURLLoaderFactory to avoid some of this identical code?

Reply:
I guess that would cause a problem for android

Reply:
yeah, because of deps on extensions :-/